File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/say/say.js
//
// This is only a SKELETON file for the 'Say' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const say = (n) => {
  if (n < 0 || n > 999999999999) {
    throw new Error('Number must be between 0 and 999,999,999,999.');
  }

  if (n === 0) return 'zero';

  const ones = [
    '', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine',
    'ten', 'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen',
    'seventeen', 'eighteen', 'nineteen'
  ];
  
  const tens = [
    '', '', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety'
  ];

  const scales = ['', 'thousand', 'million', 'billion'];

  function sayUnder100(num) {
    if (num === 0) return '';
    if (num < 20) return ones[num];
    
    const tensValue = Math.floor(num / 10);
    const remainder = num % 10;
    return remainder === 0 ? tens[tensValue] : `${tens[tensValue]}-${ones[remainder]}`;
  }

  function sayUnder1000(num) {
    if (num === 0) return '';
    
    const hundreds = Math.floor(num / 100);
    const remainder = num % 100;
    
    let result = '';
    if (hundreds > 0) {
      result = `${ones[hundreds]} hundred`;
      if (remainder > 0) result += ' ';
    }
    
    if (remainder > 0) {
      result += sayUnder100(remainder);
    }
    
    return result;
  }

  // Split number into groups of three digits
  const chunks = [];
  let remaining = n;
  while (remaining > 0) {
    chunks.push(remaining % 1000);
    remaining = Math.floor(remaining / 1000);
  }

  // Process each chunk with its scale
  const parts = chunks
    .map((chunk, index) => {
      if (chunk === 0) return '';
      const words = sayUnder1000(chunk);
      return index === 0 ? words : `${words} ${scales[index]}`;
    })
    .filter(part => part !== '')
    .reverse();

  return parts.join(' ');
};

File: src/2025-02-24-deepclaude-rust/go/exercises/practice/say/say.go
package say

import "strings"

func Say(n int64) (string, bool) {
	if n < 0 || n > 999999999999 {
		return "", false
	}
	if n == 0 {
		return "zero", true
	}

	chunks := splitChunks(n)
	scales := []string{"", "thousand", "million", "billion", "trillion"}
	var parts []string

	for i, chunk := range chunks {
		if chunk == 0 {
			continue
		}
		scaleIndex := len(chunks) - 1 - i
		chunkStr := convertLessThanThousand(chunk)
		if scaleIndex < len(scales) && scales[scaleIndex] != "" {
			chunkStr += " " + scales[scaleIndex]
		}
		parts = append(parts, chunkStr)
	}

	return strings.Join(parts, " "), true
}

func splitChunks(n int64) []int64 {
	var chunks []int64
	for n > 0 {
		chunks = append(chunks, n%1000)
		n = n / 1000
	}
	// Reverse chunks
	for i, j := 0, len(chunks)-1; i < j; i, j = i+1, j-1 {
		chunks[i], chunks[j] = chunks[j], chunks[i]
	}
	return chunks
}

func convertLessThanThousand(num int64) string {
	ones := []string{
		"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
		"ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen",
	}
	tens := []string{
		"", "ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety",
	}

	var result string

	hundreds := num / 100
	remainder := num % 100

	if hundreds > 0 {
		result += ones[hundreds] + " hundred "
	}

	if remainder > 0 {
		if remainder < 20 {
			result += ones[remainder]
		} else {
			tenPart := tens[remainder/10]
			unitPart := ones[remainder%10]
			if unitPart != "zero" {
				result += tenPart + "-" + unitPart
			} else {
				result += tenPart
			}
		}
	}

	return strings.TrimSpace(result)
}
